### PR TITLE
core: Make IRDL operations and attributes final

### PR DESCRIPTION
**Pull request by math-fehr**
It was previously possible to inherit an operation or attribute.
While this may be useful, it is also pretty dangerous, as now a constraint
such as `operand_def(MyAttribute)` will also accept other attributes such as 
`MyInheritedAttribute`, which likely shouldn't be accepted.

For instance, it was previously possible to use `riscv.fastmathflags` in the LLVM
dialect, as the attribute was inheriting `llvm.fastmathflags`.

Moreover, it is necessary for the declarative format to know that a specific attribute
type is expected, and this isn't possible if the attribute might be inherited.

The solution I'm proposing here is to use a new `final` decorator inside `irdl_attr_definition`
and `irdl_op_definition` to avoid these inheritances. Note that this `final` decorator will
trigger an error if anyone tries to subclass the final class.

Resolves: #2064

### CODE REVIEW

File: xdsl/dialects/llvm.py [hunk 1]
```diff
@@ -1170,6 +1171,14 @@ def print_parameter(self, printer: Printer):
                 )
 
 
+@irdl_attr_definition
+class FastMathAttr(FastMathAttrBase):
+    name = "llvm.fastmath"
+
+    def __init__(self, flags: None | Sequence[FastMathFlag] | Literal["none", "fast"]):
+        super().__init__(flags)
+
```
Comment (PapyChacal) [inline xdsl/dialects/llvm.py:1180]: I think you can just remove it then?

File: xdsl/dialects/riscv.py [hunk 2]
```diff
@@ -60,13 +60,17 @@
...
+class FastMathFlagsAttr(FastMathAttrBase):
     """
     riscv.fastmath is a mirror of LLVMs fastmath flags.
     """
 
     name = "riscv.fastmath"
 
+    def __init__(self, flags: None | Sequence[FastMathFlag] | Literal["none", "fast"]):
+        super().__init__(flags)
+
```
Comment (PapyChacal) [inline xdsl/dialects/riscv.py:73]: Idem

**Comment by PapyChacal** [comment]
Why reimplement `@final`?

File: xdsl/dialects/llvm.py (hunk 1 above)
Comment (math-fehr) [reply xdsl/dialects/llvm.py:1180]: Actually, no, but I don't know why ^^'
My guess is that I'm doing something wrong in the `irdl_op_definition`, but I'm not really sure why?
I'm adding an issue on this though, as this *should* work

**Comment by math-fehr** [comment]
I renamed it to `runtime_final`.
The idea is that `final` is just type annotations in Python, and do not check anything at runtime.
This is a problem for us though, as we do want to have that check at runtime, as we cannot force the typing to mark `irdl_op_definition` finals.

File: xdsl/dialects/llvm.py (hunk 1 above)
Comment (superlopuh) [reply xdsl/dialects/llvm.py:1180]: I left a comment on the issue, I think `Data.new` always calls init on `Data`, not on the class

Comment (superlopuh) [reply xdsl/dialects/llvm.py:1180]: Feels like we might as well fix that before merging this

Comment (math-fehr) [reply xdsl/dialects/llvm.py:1180]: I would suggest to not PR creep this here.
The solution is a bit more complex than expected.

The issue is that irdl_op_definition generates the `__init__`, as someone might inherit an `ABC` `Attribute` and then add more fields, so we cannot use the `super().__init__` by default. We would need to check that there is a previous `__init__`, and that no more fields are added in the child class.
So I would like to not do that here, and do this later when this become relevant (especially if we change the way we handle attributes)

Comment (superlopuh) [reply xdsl/dialects/llvm.py:1180]: fair enough

Comment (superlopuh) [reply xdsl/dialects/llvm.py:1180]: We should probably try and add some docs to explain the workaround

**Review by superlopuh — APPROVED** [review]

**Review by PapyChacal — APPROVED** [review]

**Pull request merged**